### ISSUE TITLE
[metadata] Improve validation logic for adding/updating device profiles [V2]

### ISSUE THREAD

**Issue by tonyespy** [edgexfoundry/edgex-go#2647]
# 🐞 Bug Report

### Affected Services
 The issue is located in: Metadata

### Is this a regression?

No

### Description and Minimal Reproduction
It's possible to add new device profiles to the system which are not valid, for instance a device profile which only defines name, description, and labels can be added.  Ex.

```
curl -w %{http_code} -X POST \
-d '
name: "Device-FOO"
manufacturer: "Acme"
model: "FOO"
labels:
- "device-foo-example"
description: "Example of Device-FOO"
' http://localhost:48081/api/v1/deviceprofile/upload
```

Likewise, it's also possible to add a device profile with invalid objects (i.e. device resources, device commands, or core commands) which then cannot actually be used. Our existing core-contracts model-based validation logic does basic checking, for instance it ensures that a device command must specify a non-empty **deviceResource** or **object** attribute, however it doesn't validate that the **deviceResource** or **object** actually references an existing **deviceResource**. Here's another example where a device profile defines a single device resource and single device command, however the device command references a non-existent device resource.  Ex.

```
curl -w %{http_code} -X POST \
-d '
name: "Roadrunner Device Profile"
manufacturer: "Acme"
model: "Roadrunner"
labels:
- "roadrunner"
description: "Roadrunner example device profile"

deviceResources:
-
  name: "Roadrunner"
  description: "indicate whether the Roadrunner is on the road"
  properties:
    value:
      { type: "Bool", readWrite: "W", defaultValue: "true" }
    units:
      { type: "String", readWrite: "R", defaultValue: "Random" }

deviceCommands:
-
  name: "GetRoadRunner"
  get:
  - { operation: "get", deviceResource: "Bool", object="Bool"}
' http://localhost:48081/api/v1/deviceprofile/upload

```

Some outstanding questions (some which may be answered in our Device Services SDK requirements document) should be answered before we start on any sort of implementation:

 *  what should a minimum viable device profile contain (i.e. a single device resource)?
 * should we continue to support both **deviceResource** and **object** for device commands?
 * can the core command section of a device profile be auto-generated? It seems to me that many of the status code return values being described are really standard responses across all device services. Likewise, the core commands just map to device commands (and/or device resources?) which leads me to think that this section could be auto-generated.

## 🔥 Exception or Error
<pre><code>
 This depends on the contents of the imported device profile, in at least one known case (issue #2596) a bad profile can cause the deviceprofile endpoint to always return an error on queries or deletes.

Mostly though, the end result is a device profile that's not really usable, and just taking up space in the db.

</code></pre>

## 🌍  Your Environment
**Deployment Environment:**
All

**EdgeX Version:**
Geneva

**Anything else relevant?**

**Comment by cloudxxx8** [comment]
may we consider this improvement in V2, and let V1 as is?
add @weichou1229 for awareness

**Comment by lenny-intel** [comment]
> may we consider this improvement in V2, and let V1 as is?

@cloudxxx8 , Yes. Let make this a V2 issue. I comment on the DTO PR that it sill needs more robust validation. We can use this issue to address it so we can move forward with the V2 implementation.

**Comment by jpwhitemn** [comment]
Needs to be part of V2 API and before EdgeX 2.0.  May have to slide this to Ireland.

**Comment by weichou1229** [comment]
For V2 implementation,

> what should a minimum viable device profile contain (i.e. a single device resource)?

We need to define the name and at least one of the device resources. The definition can refer to DTO https://github.com/edgexfoundry/go-mod-core-contracts/blob/master/v2/dtos/deviceprofile.go#L17


> should we continue to support both deviceResource and object for device commands?

> can the core command section of a device profile be auto-generated? It seems to me that many of the status code return values being described are really standard responses across all device services. Likewise, the core commands just map to device commands (and/or device resources?) which leads me to think that this section could be auto-generated.

We simplify the core-command, the user just need to enable the Get or Put operation. The definition can refer to https://github.com/edgexfoundry/go-mod-core-contracts/blob/master/v2/dtos/command.go#L13

**Comment by jpwhitemn** [comment]
Ireland Planning Meeting stretch goal

**Comment by cloudxxx8** [comment]
@tonyespy @iain-anderson our users encounter some strange behavior when they create the problematic Device Profile, so I would like to implement this as early as possible.  Aside from the rules written by Bruce, could you please help add other validation rules in your mind?

**Comment by tonyespy** [comment]
@cloudxxx8 I think the rules from Bruce cover most cases. The only other rule that we probably should implement is:

 * deviceResources referenced in deviceComands must exist

Let's start with this set of rules, and if during this cycle we discover other issues, we can extend them.

**Comment by cloudxxx8** [comment]
@jinlinGuan for testing awareness

**Comment by cloudxxx8** [comment]
we've implemented the validation rules stated in this issue in https://github.com/edgexfoundry/go-mod-core-contracts/pull/396
